File: 0817-Linked List Components/0817-Linked List Components.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public int numComponents(ListNode head, int[] G) {
        Set<Integer> table = new HashSet<>();
        for (int i : G) {
            table.add(i);
        }
        ListNode curr = head;
        int count = 0;
        while (curr != null) {
            if (table.contains(curr.val) && (curr.next == null || !table.contains(curr.next.val))) {
                ++count;
            }
            curr = curr.next;
        }
        return count;
    }
}

File: 0817-Linked List Components/0817-Linked List Components.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def numComponents(self, head, G):
        """
        :type head: ListNode
        :type G: List[int]
        :rtype: int
        """
        table = set(G)
        curr = head
        count = 0
        while curr:
            if curr.val in table and (curr.next == None or curr.next.val not in table):
                count += 1
            curr = curr.next
        return count
